src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -595,4 +612,216 @@ public static string GetTempPath(string workingDir)
             return Environment.GetEnvironmentVariable("SYSTEMROOT");
         }
     }
+
+    internal interface IServerMutex : IDisposable
+    {
+        bool TryLock(int timeoutMs);
+        void Unlock();
+        bool IsDisposed { get; }
+    }
+
+    internal sealed class FileMutex : IDisposable
+    {
+        public readonly bool OwnsMutex;
+        public readonly FileStream Stream;
+        public readonly string FilePath;
+
+        public bool IsLocked { get; private set; }
+
+        internal static string GetMutexDirectory()
+        {
+            var homeDirectory = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
+            if (String.IsNullOrWhiteSpace(homeDirectory))
+                homeDirectory = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
+            if (String.IsNullOrWhiteSpace(homeDirectory))
+                homeDirectory = Path.GetTempPath();
+            var result = Path.Combine(homeDirectory, ".roslyn");
+            if (!Directory.Exists(result))
+                Directory.CreateDirectory(result);
+            return result;
+        }
+
+        public FileMutex(string name)
+        {
+            FilePath = Path.Combine(GetMutexDirectory(), name);
+            try
+            {
+                Stream = new FileStream(FilePath, FileMode.Open, FileAccess.ReadWrite, FileShare.None);
+                OwnsMutex = false;
+            }
+            catch (FileNotFoundException)
+            {
+                Stream = new FileStream(FilePath, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.None);
+                OwnsMutex = true;
+            }
+        }
```

Comment:
What invariant is this function trying to maintain? It's not immediately clear what `OwnsMutex` means here.

Reply:
I think it's left over from before I realized you can't delete the file. I'll remove it.